plugins/catalog-backend-module-ldap/src/ldap/client.ts

```diff
@@ -147,39 +147,57 @@ export class LdapClient {
   async searchStreaming(
     dn: string,
     options: SearchOptions,
-    f: (entry: SearchEntry) => void,
+    f: (entry: SearchEntry) => Promise<void> | void,
   ): Promise<void> {
     try {
+      let awaitList: Array<Promise<void> | void> = [];
       return await new Promise<void>((resolve, reject) => {
         // Note that we clone the (frozen) options, since ldapjs rudely tries to
         // overwrite parts of them
-        this.client.search(dn, cloneDeep(options), (err, res) => {
-          if (err) {
-            reject(new Error(errorString(err)));
-          }
-
-          res.on('searchReference', () => {
-            this.logger.warn('Received unsupported search referral');
-          });
-
-          res.on('searchEntry', entry => {
-            f(entry);
-          });
-
-          res.on('error', e => {
-            reject(new Error(errorString(e)));
-          });
-
-          res.on('end', r => {
-            if (!r) {
-              throw new Error('Null response');
-            } else if (r.status !== 0) {
-              throw new Error(`Got status ${r.status}: ${r.errorMessage}`);
-            } else {
-              resolve();
+        this.client.search(
+          dn,
+          merge(cloneDeep(options), { paged: { pagePause: true } }),
+          (err, res) => {
+            if (err) {
+              reject(new Error(errorString(err)));
             }
-          });
-        });
+
+            res.on('searchReference', () => {
+              this.logger.warn('Received unsupported search referral');
+            });
+
+            res.on('searchEntry', entry => {
+              awaitList.push(f(entry));
+            });
+
+            res.on('page', (_, cb) => {
+              // awaits completion before fetching next page
+              Promise.all(awaitList)
+                .then(() => {
+                  // flush list
+                  awaitList = [];
+                  if (cb) cb();
+                })
+                .catch(reject);
```

Comment:
I think that even the error handler needs to call the cb, right? I'm worried the client might be left hanging indefinitely if we don't. Is there some client cancellation thing we can run at that point too, to cut off the rest of the fetching?

Reply:
There isn't a client cancellation function, I think we can do a `Promise.allSettled` so we don't reject when we have an error in the transform function, in this way, we only reject if we have a connection error.